Make applying DeepReadonly idempotent

aaronjensen opened the pull request:
fixes #106

This isn't as clean as I'd like—ideally a recursive application would have no impact. For example, the newly added test would match as if it weren't applied twice:

```
exports[`DeepReadonly testType<DeepReadonly<DeepReadonly<NestedArrayProps>>['first']['second']>() 1`] = `"_DeepReadonlyArray<_DeepReadonlyObject<{ name: string; }>>"`;
```

Above, the `_DeepReadonlyObject` is superfluous and isn't there if applied only once.

```
exports[`DeepReadonly testType<DeepReadonly<NestedArrayProps>['first']['second']>() 1`] = `"_DeepReadonlyArray<{ name: string; }>"`;
```

<!-- Thank you for your contribution! :thumbsup: -->
<!-- Please makes sure that these checkboxes are checked before submitting your PR, thank you! -->

## Description
<!-- Example: Added error property support to `action` API -->

## Related issues:
- Resolved #106

## Checklist

* [x] I have read [CONTRIBUTING.md](../CONTRIBUTING.md)
* [x] I have linked all related issues above
* [x] I have rebased my branch

For bugfixes:
* [x] I have added at least one unit test to confirm the bug have been fixed
* [x] I have checked and updated TOC and API Docs when necessary

For new features:
* [x] I have added entry in TOC and API Docs
* [x] I have added a short example in API Docs to demonstrate new usage
* [x] I have added type unit tests with `dts-jest`

aaronjensen commented:
Actually, it turns out that the issue I pointed to above is a real problem, those are two different types according to typescript... Any ideas on how to detect that we've already applied `DeepReadonly` and short-circuit?

steida commented:
Maybe this will help

(More) Recursive Type Aliases
https://devblogs.microsoft.com/typescript/announcing-typescript-3-7-beta/

aaronjensen commented:
Okay, I've pushed up a commit with a new solution that uses a symbol tag to skip applying DeepReadonly again. I'm not sure how you feel about privateish symbol tags, but this is the only way I know how to do something like this.

aaronjensen commented:
@steida I think that'll help with the ultimate definition so we don't have to create separate interfaces to enable recursion, but I'm not sure it helps with this directly, but maybe I'm missing something.

aaronjensen commented:
Unfortunately, the tagging taints Records, which makes things like `Object.entries` not work like you'd want it to. I'm currently at a loss on how to express this properly and make it idempotent.

aaronjensen commented:
Okay, 24d4555 seems to work like I want it to so far. The only downside is it strips the names of types that are made ReadOnly ([see here](https://github.com/piotrwitek/utility-types/pull/107/files#diff-168cbee1e98ceb5c1c644b9c771bea02R31)).

piotrwitek commented on src/mapped-types.ts:
```diff
@@ -412,12 +412,18 @@ export type PromiseType<T extends Promise<any>> = T extends Promise<infer U>
+  : T extends _DeepReadonlyArray<infer U>
+  ? _DeepReadonlyArray<U>
+  : T extends _DeepReadonlyObject<infer V>
+  ? _DeepReadonlyObject<V>
```
Can't we flip that so it will first match _DeepReadonlyArray & _DeepReadonlyObject? Then the rest seems to be always mapped to T anyway.
It would be much cleaner.

aaronjensen replied:
If I understand what you're asking correctly, then no, `string` and `boolean` actually match to `_DeepReadonlyObject<infer V>`. It extracts out things like `toString()` and whatnot.

piotrwitek replied:
Ok got it they need to short circuit before them.

piotrwitek commented on src/mapped-types.ts:
```diff
@@ -412,12 +412,18 @@ export type PromiseType<T extends Promise<any>> = T extends Promise<infer U>
+  | number
+  | boolean
+  | null
+  | undefined
```
What is the rationale behind null and undefined here? If there is something please add test cases to cover it.

aaronjensen replied:
Completeness, doesn't look necessary (added a test) so I removed it.

piotrwitek commented on src/mapped-types.ts:
```diff
@@ -412,12 +412,18 @@ export type PromiseType<T extends Promise<any>> = T extends Promise<infer U>
+  | ((...args: any[]) => any)
+  | string
+  | number
+  | boolean
```
I would replace it with Primitive type, it will guard against other edge cases

aaronjensen replied:
Didn't realize there was one, perfect.

piotrwitek commented on src/mapped-types.spec.ts:
```diff
@@ -315,6 +315,8 @@ type RequiredOptionalProps = {
   // @dts-jest:pass:snap
   testType<DeepReadonly<NestedProps>['first']>();
   // @dts-jest:pass:snap
+  testType<DeepReadonly<DeepReadonly<NestedProps>['first']>>();
```
Please don't add tests randomly into nicely ordered tests cases, add new ones at the end of the logical block

aaronjensen replied:
got it, I put them in their own section, thanks.

piotrwitek commented on src/mapped-types.spec.ts:
```diff
@@ -329,6 +331,8 @@ type RequiredOptionalProps = {
   // @dts-jest:pass:snap
   testType<DeepReadonly<NestedArrayProps>['first']['second']>();
   // @dts-jest:pass:snap
+  testType<DeepReadonly<DeepReadonly<NestedArrayProps>>['first']['second']>();
```
ditto

aaronjensen commented:
Addressed feedback in 9fa8988, I think there may be a conflict between the dts snap and prettier now, however—not sure how you want to resolve that. I could split up the primitive test so prettier doesn't want to reformat. `ci-check` does pass, so maybe it's fine.

aaronjensen commented:
Also, let me know if you'd like me to squash the commits or if you just plan on doing it w/ GitHub. Thanks!

piotrwitek commented on src/mapped-types.spec.snap.ts:
```diff
@@ -342,6 +342,16 @@ type RequiredOptionalProps = {
+  // @dts-jest:pass:snap -> string | number | bigint | boolean | symbol | null | undefined
+  testType<
+    DeepReadonly<string | null | undefined | boolean | number | bigint | symbol>
+  >();
```
Looks good 👍

piotrwitek commented on src/mapped-types.ts:
```diff
@@ -412,12 +412,12 @@ export type PromiseType<T extends Promise<any>> = T extends Promise<infer U>
+  : T extends _DeepReadonlyArray<infer U>
+  ? _DeepReadonlyArray<U>
+  : T extends _DeepReadonlyObject<infer V>
+  ? _DeepReadonlyObject<V>
```
Looks good!
